iwyu.cc

```diff
@@ -2712,7 +2646,83 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // void fn(const S<Class>& s) { // Forward declarations are sufficient here.
   //   (void)s.t; // Full 'Class' type is needed due to template instantiation.
   // }
-  void ReportTplSpecComponentTypes(const TemplateSpecializationType*) = delete;
+  void ReportTplSpecComponentTypes(const TemplateSpecializationType*,
+                                   const set<const Type*>& blocked_types) =
+      delete;
+
+  const std::set<const Type*>& GetBlockedTypes() const = delete;
+
+  void ReportTypeUseInternal(SourceLocation used_loc, const Type* type,
+                             const char* comment,
+                             const set<const Type*>& blocked_types) {
```

Comment:
I'm still not 100% on the "blocked" terminology.

I see that the callers use it to filter out already-provided types, and I appreciate the layering that we could use it for something else as well, i.e. we probably don't want it to be called `provided_types` here.

Or would that be clearer? Less general, for sure, but more to the point.

Reply:
This function reports `type` and its underlying types without types listed in `blocked_types`. So, this parameter is exactly for that purpose as it is named — it prevents some types from reporting — except the case with type aliases when a set of types "provided" with typedef supersedes `blocked_types` set.

Maybe, you have some suggestions about naming?

Reply:
The best I can come up with is `provided_types`, but I agree that that is not ideal.

It would be nice if it was possible to invert the logic entirely -- recursively collect all types we want to report starting from a `Type*`, filter out already-provided ("blocked") types, and then just call a dumber `ReportTypesUse` that simply does the reporting without any smarts for type components.

In general I like it when logic flows strictly forwards instead of like here, looking back into some set we prepared earlier to see if we should stop working. Hard to explain, but it feels like we're jumping back and forth here.

Reply:
Unluckily, `ReportTypeUse` cannot simply report a type because it should itself launch `InstantiatedTemplateVisitor` analysis if the type occurs to be `TemplateSpecializationType` (see [`ReportTplSpecComponentTypes`](https://github.com/include-what-you-use/include-what-you-use/blob/0.20/iwyu.cc#L1688)). `blocked_types` parameter is mostly for propagating into it. In principle, it can be inverted into `allowed_types` along with the field member of `InstantiatedTemplateVisitor`, but I think that `blocked_types` suits better to [proposed implementation of arbitrary dependent type reporting](https://github.com/include-what-you-use/include-what-you-use/pull/1182#issuecomment-1406681708):
> A solution to this may be to traverse an original template declaration with "reporting" types in a list, and then use this list as a list of types blocked for reporting on template specialization decl traversal. Reports of all other types should then be allowed.

Reply:
> It would be nice if it was possible to invert the logic entirely -- recursively collect all types we want to report starting
> from a Type*, filter out already-provided ("blocked") types, and then just call a dumber ReportTypesUse that simply 
> does the reporting without any smarts for type components.

I realize it's not super useful in its current form, but I wanted to sketch out what I meant above:
https://github.com/kimgr/include-what-you-use/commit/kg-expand-instead-of-report

It feels like we could replace InstantiatedTemplateVisitor with something that just collects types, but maybe I'm just romanticizing :)

Reply:
Just to be clear: I don't think it's realistic to adopt any of my sketch above, particularly not for this PR, but I'm interested to hear what you think of its viability. Would it be possible to decouple `ReportTplSpecComponentTypes` from the AST visitation?

Reply:
I think it may be just a naming problem. I. e., `ReportTypeUse` is not actually `Report...` but `Handle...` or something like that?

Btw, "reporting" types to `std::set` should probably be done on implementing improvement of template dependent type handling.

> There are never any blocked types/provided types in IwyuAstConsumer

It will probably not be the case after implementing separate type component handling for fn-return and "autocast".

Reply:
I understand your wish to diminish complexity, but `ReportTypeUse` is currently the central place to handle types that are needed fully but aren't explicitly written in the source. If that type occurs to be typedef, the "provision" logic should show whether its underlying type is needed or not. If that type occurs to be a template specialization, the handling should be done similar to as if it were explicitly written, and currently `InstantiatedTemplateVisitor` performs that handling.

Reply:
> Would it be possible to decouple ReportTplSpecComponentTypes from the AST visitation?

I think recursive calls `Traverse... -> Report(Handle)TypeUse -> Traverse... -> Report(Handle)TypeUse -> ...` are unavoidable in general. Consider:
```cpp
struct A {};

template <typename T>
struct Tpl3 {
    int b;
    T t;
};

template <typename T>
struct Tpl2 {
    int a;
    Tpl3<T>* t3;
    decltype(t3->b) n;
};

template <typename T>
struct Tpl1 {
    int i;
    Tpl2<T>* t2;
    decltype(t2->a) k;
};

void Fn(Tpl1<A>* t1) {
    (void)sizeof(*t1);
}
```
`sizeof` expression in `Fn` constitutes full use of `Tpl1<A>`, which requires full type info for its member fields, whereas declaration of one of them (`Tpl1::k`) contains `MemberExpr` for `Tpl2<A>`, constituting its full use, which requires full type info for its fields, and so on. (Currently, IWYU doesn't handle this case correctly because `CanIgnoreType` with default argument `IgnoreKind::ForUse` returns `true` for `Tpl2<A>` during `Tpl1<A>` visitation).

But if you insist, I can try to implement "reporting" into `std::set` on an experimental branch. `Report(Expand)TplSpecComponentTypes` would then use its own stack-allocated `InstantiatedTemplateVisitor` object. This approach should allow using `set_difference` instead of passing `blocked_types` into `InstantiatedTemplateVisitor`, as you've suggested.

Reply:
> But if you insist, I can try to implement "reporting" into std::set on an experimental branch.
> Report(Expand)TplSpecComponentTypes would then use its own stack-allocated
> InstantiatedTemplateVisitor object. This approach should allow using set_difference instead
> of passing blocked_types into InstantiatedTemplateVisitor, as you've suggested.

I'm in no position to insist :), but it would be interesting to see. As I've mentioned several times, my understanding of the template traversal is fuzzy at best, so I don't feel comfortable trying to hack it up myself.

I don't think we should block this PR on that experiment, however. I would like to get this merged, but I still find the `blocked_types` terminology a little off. Will try to find time to come up with other suggestions.